src/vsock/src/packet.rs

```diff
@@ -0,0 +1,344 @@
+#![deny(missing_docs)]
+
+use super::vhu_vsock::{Error, Result};
+use byteorder::{ByteOrder, LittleEndian};
+use virtio_queue::DescriptorChain;
+use vm_memory::{
+    GuestAddress, GuestAddressSpace, GuestMemory, GuestMemoryAtomic, GuestMemoryLoadGuard,
+    GuestMemoryMmap,
+};
+
+// TODO: Replace below with bindgen generated struct
+// vsock packet header size when packed
+const VSOCK_PKT_HDR_SIZE: usize = 44;
+
+// Offset into header for source cid
+const HDROFF_SRC_CID: usize = 0;
+
+// Offset into header for destination cid
+const HDROFF_DST_CID: usize = 8;
+
+// Offset into header for source port
+const HDROFF_SRC_PORT: usize = 16;
+
+// Offset into header for destination port
+const HDROFF_DST_PORT: usize = 20;
+
+// Offset into the header for data length
+const HDROFF_LEN: usize = 24;
+
+// Offset into header for packet type
+const HDROFF_TYPE: usize = 28;
+
+// Offset into header for operation kind
+const HDROFF_OP: usize = 30;
+
+// Offset into header for additional flags
+// only for VSOCK_OP_SHUTDOWN
+const HDROFF_FLAGS: usize = 32;
+
+// Offset into header for tx buf alloc
+const HDROFF_BUF_ALLOC: usize = 36;
+
+// Offset into header for forward count
+const HDROFF_FWD_CNT: usize = 40;
+
+/// Vsock packet structure implemented as a wrapper around a virtq descriptor chain:
+/// - chain head holds the packet header
+/// - optional data descriptor, only present for data packets (VSOCK_OP_RW)
+pub struct VsockPacket {
+    hdr: *mut u8,
+    buf: Option<*mut u8>,
+    buf_size: usize,
+}
+
+impl VsockPacket {
+    /// Create a vsock packet wrapper around a chain in the rx virtqueue.
+    /// Perform bounds checking before creating the wrapper.
+    pub fn from_rx_virtq_head(
+        chain: &mut DescriptorChain<GuestMemoryAtomic<GuestMemoryMmap>>,
+        mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    ) -> Result<Self> {
+        // head is at 0, next is at 1, max of two descriptors
+        // head contains the packet header
+        // next contains the optional packet data
+        let mut descr_vec = Vec::with_capacity(2);
+
+        let mut num_descr = 0;
+        loop {
+            if num_descr >= 2 {
+                // Maybe this should be an error?
+                break;
+            }
+            let descr = chain.next().ok_or(Error::QueueMissingDescriptor)?;
+            if !descr.is_write_only() {
+                return Err(Error::UnwritableDescriptor);
+            }
+            num_descr += 1;
+            descr_vec.push(descr);
+        }
+
+        let head_descr = descr_vec[0];
+        let data_descr = descr_vec[1];
```

Comment:
in fact a match descr_vec.len() might be better because you can error out if you don't have exactly 2 descriptors which the code here assumes.